File: java/bank-account/src/main/java/BankAccount.java
public class BankAccount {

    private int balance;
    private boolean isOpen;

    public BankAccount() {
    }

    public void open() {
        isOpen = true;
        balance = 0;
    }

    public int getBalance() throws BankAccountActionInvalidException {

        if (!isOpen) { throw new BankAccountActionInvalidException("Account closed"); }
        return balance;
    }

    public synchronized void deposit(int amount) throws BankAccountActionInvalidException {

        if (!isOpen)    { throw new BankAccountActionInvalidException("Account closed"); }
        if (amount < 0) { throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount"); }

        balance += amount;
    }

    public synchronized void withdraw(int amount) throws BankAccountActionInvalidException {

        if (!isOpen)           { throw new BankAccountActionInvalidException("Account closed"); }
        if (balance == 0)      { throw new BankAccountActionInvalidException("Cannot withdraw money from an empty account"); }
        if (amount > balance ) { throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account"); }
        if (amount < 0)        { throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount"); }

        balance -= amount;
    }

    public synchronized void close() {
        isOpen = false;
    }
}

File: javascript/bank-account/bank-account.js
"use strict";

export class BankAccount {

  constructor() {
    this.isOpen = false;
    this.currentBalance = 0;
  }

  open() {

    if (this.isOpen) {
      throw new ValueError();
    } else {
      this.isOpen = true;
    }
  }

  close() {

    if (this.isOpen) {
      this.isOpen = false;
      this.currentBalance = 0;
    } else {
      throw new ValueError();
    }
  }

  deposit(amount) {

    if (!this.isOpen || amount < 0) {
      throw new ValueError();
    }

    this.currentBalance += amount;
  }

  withdraw(amount) {

    if (!this.isOpen || this.currentBalance == 0 || amount > this.currentBalance || amount < 0 ) {
      throw new ValueError();
    }

    this.currentBalance -= amount;
  }

  get balance() {

    if (!this.isOpen) {
      throw new ValueError();
    }

    return this.currentBalance;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
